Add base django app

jmaupetit opened the pull request:
## Purpose

We need some preliminary work to build a comfortable development environment to integrate the xAPI library in openedx.

## Proposal

- [x] add base Django application
- [x] add application packaging
- [x] add docker environment (Dockerfile + docker-compose) based on openedx's hawthorn release
- [x] add learning locker LRS services
- [x] update documentation (README)

jmaupetit committed:
🏗(core) add base xapi django app

Add a raw Django application as generated by the django-admin startapp
command.

jmaupetit committed:
📦(core) add project packaging

Add base project packaging configuration.

Note: setuptools seems to have issues when dealing with projects not
located in the project's root [1]. Let's be pragmatic and remove the
src/ directory.

1. https://github.com/pypa/setuptools/issues/1136

jmaupetit committed:
🐳(development) add base configuration for dev

Add base docker environment to develop the xapi app in the context of
openedx LMS (hawthorn).

Note that we override LMS settings during the build in the following file:

docker/config/lms/docker_run_development.py

You will need to rebuild your image if you change those settings. If
settings should be more versatile during development, one should mount
them as a docker volume.

jmaupetit committed:
🐳(cmd) install and use dockerize

When running "make bootstrap" for the first time, starting the mysql
container can take some time before it gets fully operationnal. We use
dockerize to wait for mysql to be up and running before trying to start
the LMS django development server and perform database migrations.

jmaupetit committed:
🐳(services) integrate learning locker services

Following up2university's learning locker integration [1], we've added
the following services to our development stack:

- learning locker api
- learning locker ui
- learning locker xapi (the lms depends on it)
- learning locker worker (unused for now)
- redis
- nginx (reverse proxy to the ui & api services)

1. https://github.com/up2university/learninglocker2

jmaupetit committed:
📝(README) add base projet bootstrapping docs

This is a draft basic documentation to ease project bootstrapping by
developer willing to contribute.

jmaupetit commented on Dockerfile:
```diff
@@ -30,6 +30,8 @@ RUN pip install --no-cache-dir --src /usr/local/src -e /edx/app/xapi
     cp /edx/app/edxapp/edx-platform/common/static/webpack-stats.json /edx/app/edxapp/staticfiles/
 
+# Install dockerize to wait for mysql before running the container command
+# (and prevent connection issues)
```
I know @sampaccoud would have been unhappy without this comment :wink:

sampaccoud replied:
:thinking:

jmaupetit replied:
Just kidding, because I forgot this comment, and I knew you would told me to add it.

sampaccoud commented on .env:
```diff
@@ -0,0 +1,48 @@
+CELERY_ALWAYS_EAGER=1
+
+# -- Features --
+FEATURES={"AUTOMATIC_AUTH_FOR_TESTING": true, "RESTRICT_AUTOMATIC_AUTH": false}
```
Maybe add one line to explain what this unlocks?

jmaupetit replied:
Done.

sampaccoud commented on Dockerfile:
```diff
@@ -0,0 +1,42 @@
+# webpack-stats.json is required both in production and development in a static
+# directory 😢
+RUN mkdir /edx/app/edxapp/staticfiles && \
+    cp /edx/app/edxapp/edx-platform/common/static/webpack-stats.json /edx/app/edxapp/staticfiles/
```
but this is already done in the image you inherit from?

jmaupetit replied:
Unfortunately no: this file is available in the `builder` and `development` targets, but not in `production` where you are supposed to perform a `collectstatic`.

sampaccoud replied:
But if I'm not mistaken edx originally puts this file in `/edx/app/edxapp/staticfiles/` and we moved it to `/edx/app/edxapp/edx-platform/common/static/webpack-stats.json`.
So you are doing the other way around. Why not make a copy in the fun-platform image and keep it in both places?

jmaupetit replied:
I'm not comfortable with this as the production image is not supposed to have "collected" static files... which will be overriden _via_ a docker volume in 90% cases. I might be wrong but I think it's best to fix horrible things in derived images than in the production one. WDYT?

sampaccoud commented on Dockerfile:
```diff
@@ -0,0 +1,42 @@
+
+# Add the non-privileged user that will run the application
+RUN groupadd --gid ${GID} edx && \
+    useradd --uid ${UID} --gid ${GID} --home /edx edx
```
Why do we need to do this and not in fun-platform?

If necessary, edx is not a good name for this, why not use edxapp?

jmaupetit replied:
We are running this image only for local development purpose (and not production + development). This image is not supposed to be working in OpenShift or other platforms that assign a random non-privileged user ID to run a container. Hence, when working with this image, it's more **comfortable** to rely on an existing user (when you log in the container, etc.) than a random one activated _via_ an entrypoint (that is therefore not required here).

Moreover (even if it's not required in our case), the `COPY --chown` docker statement requires an existing system user (you cannot use the entrypoint trick during the build).

Concerning the group/username, `edx:edx` seemed relevant to me, but `edxapp` is also fine.

sampaccoud replied:
ok! I think this explanation needs to go somewhere in the code!

jmaupetit replied:
> ok! I think this explanation needs to go somewhere in the code!

:thinking: Sure? 

> why not use edxapp?

Should I switch to it?

sampaccoud replied:
edxapp for user: yes it's the common openedx practice.

sampaccoud replied:
> :thinking:  Sure?
I meant in the repo... doc?

jmaupetit replied:
I've added a header mentioning this in the `Dockerfile`.

jmaupetit committed:
📄(update) move from GPLv3 to AGPLv3

AGPLv3 license seems to be the norm for openedx-related projects. Let's
stick to it.
